utils/git/code-format-diff-apply.py

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python3
+#
+# ====- code-format-diff-apply, apply diff from comment --*- python -*-------==#
+#
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# ==-------------------------------------------------------------------------==#
+
+import argparse
+import os
+import re
+import subprocess
+import sys
+import tempfile
+from functools import cached_property
+
+import github
+from github import IssueComment, PullRequest
+
+LF = '\n'
+CRLF = '\r\n'
+CR = '\r'
+
+diff_pat = re.compile(r"``````````diff(?P<DIFF>.+)``````````", re.DOTALL)
+
+def get_diff_from_comment(comment: IssueComment.IssueComment) -> str:
+    m = re.search(diff_pat, comment.body)
+    if m is None:
+        raise Exception(f"Could not find diff in comment {comment.id}")
+    diff = m.group("DIFF")
+    # force to linux line endings
+    diff = diff.replace(CRLF, LF).replace(CR, LF)
+    return diff
+
+def run_cmd(cmd: [str]) -> None:
+    print(f"Running: {' '.join(cmd)}")
+    proc = subprocess.run(cmd, capture_output=True, text=True)
+    if proc.returncode != 0:
+        print(proc.stdout)
+        print(proc.stderr)
+        raise Exception(f"Failed to run {' '.join(cmd)}")
+
+def apply_patches(args: argparse.Namespace) -> None:
+    repo = github.Github(args.token).get_repo(args.repo)
+    pr = repo.get_issue(args.issue_number).as_pull_request()
+
+    comment = pr.get_issue_comment(args.comment_id)
+    if comment is None:
+        raise Exception(f"Comment {args.comment_id} does not exist")
+
+    # get the diff from the comment
+    diff = get_diff_from_comment(comment)
+
+    # write diff to temporary file and apply
+    with tempfile.NamedTemporaryFile() as tmp:
+        tmp.write(diff.encode("utf-8"))
+        tmp.flush()
+
+        # run git apply tmp.name
+        apply_cmd = [
+            "git",
+            "apply",
+            tmp.name
+        ]
+        run_cmd(apply_cmd)
+
+    # run git add .
+    add_cmd = [
```

Comment:
One thing I ran into when manually clang formatting was outdated SPIR-V headers. In fact, I think it is quite common  to accidentally add modified SPIR-V headers through git add . when it's not intended. Is there a way to loop through the actual diff text and collect all the files that are mentioned in the diff, rather than running git add .? Maybe with a regex? Then we can run git add <filename> for each one.

Reply:
The build machines should checkout correctly, and in theory it shouldn't have any out-of-date spir-v headers or any other files that may be floating around that weren't impacted by clang-format, so my concern is resolved.